src/gfw/common/beam/transforms/read_matching_avro_files.py

```diff
@@ -0,0 +1,180 @@
+
+        decode:
+            Whether to decode the data from bytes to string.
+            Default is True.
+
+        decode_method:
+            The method used to decode the message data.
+            Supported methods include standard encodings like "utf-8", "ascii", etc.
+            Default is "utf-8".
+
+        read_all_from_avro_kwargs:
+            Any additional keyword arguments to be passed to Beam's `ReadAllFromAvro` class.
+            Check official documentation:
+            https://beam.apache.org/releases/pydoc/2.64.0/apache_beam.io.avroio.html#apache_beam.io.avroio.ReadAllFromAvro
+
+        **kwargs:
+            Additional keyword arguments passed to base PTransform class.
+
+    Returns:
+        PCollection:
+            A PCollection of Avro records from the files within the specified datetime range.
+    """
+
+    DATETIME_REGEX = r"(\d{4}-\d{2}-\d{2}).*?(\d{2}_\d{2}_\d{2}Z)"
+    PATTERN_TEMPLATE = "{base_path}/{folder_path}/*.avro"
+
+    def __init__(
+        self,
+        base_path: str,
+        start_dt: str,
+        end_dt: str,
+        path_template: str = "{date}",
+        decode: bool = True,
+        decode_method: str = "utf-8",
+        read_all_from_avro_kwargs: Optional[dict[Any, Any]] = None,
+        **kwargs: Any,
+    ) -> None:
+        super().__init__(**kwargs)
+        self._base_path = base_path
+        self._start_dt = datetime_from_string(start_dt)
+        self._end_dt = datetime_from_string(end_dt)
+        self._path_template = path_template
+        self._decode = decode
+        self._decode_method = decode_method
+        self._read_all_from_avro_kwargs = read_all_from_avro_kwargs or {}
+
+        self._validate_decode_method()
+
+    def _generate_file_patterns(self) -> Sequence[str]:
+        current_date = self._start_dt.date()
+        end_date = self._end_dt.date()
+        patterns = []
+
+        while current_date <= end_date:
+            patterns.append(
+                self.PATTERN_TEMPLATE.format(
+                    base_path=self._base_path,
+                    folder_path=self._path_template.format(date=current_date.isoformat()),
+                )
+            )
+            current_date += timedelta(days=1)
+
+        return patterns
+
+    def _validate_decode_method(self) -> None:
+        try:
+            codecs.lookup(self._decode_method)
+        except LookupError as e:
+            raise ValueError(f"Unsupported decode method: {self._decode_method}") from e
+
+        logger.info(f"Using decode method: {self._decode_method}.")
+
+    def _decode_records(self, record: dict) -> dict:
+        record = {**record}
+        record["data"] = record["data"].decode(self._decode_method)
+
+        return record
+
+    def is_path_in_range(self, path: str) -> bool:
+        """Checks if a path containing a datetime is within the provided datetime range."""
+        dt = get_datetime_from_string(path)
```

Comment:
Why creating a common function to extract and parse a datetime? I think this should be divided in two steps:
1. extract the datetime string from the file: This could be extracted to a string parsing function if you like to return the  `2025-08-14_09_30_00Z` part from the fiename. With this you could use the `DATETIME_REGEX` defined above.
2. parse the datetime string into a datetime object: It can be done directly here using `datetime.strptime`

The implementation would be something like:
```
       datetime_str = extract_datetime_string(path, DATETIME_REGEX)
       dt = datetime.strptime(datetime_str, "%Y-%m-%d_%H_%M_%SZ").replace(tzinfo=timezone.utc)
```